### ISSUE TITLE
For unauthorized user error code 401 should be returned instead of 403

### ISSUE THREAD

**Issue by maxceem** [topcoder-platform/tc-project-service#442]
Currently, if we try to access Project Service without providing a token, we would get error code `403` [Forbidden](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403):

![image](https://user-images.githubusercontent.com/146016/72490617-b007db00-3852-11ea-907a-ff968b182ea8.png)

Though, as per our swagger file and per status definition we suppose to return error code `401` [Unauthorized](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401) in such cases:

![image](https://user-images.githubusercontent.com/146016/72490818-3a503f00-3853-11ea-872f-3e7e789c81d3.png)

**Comment by vikasrohit** [comment]
Good catch @maxceem. We should throw 401 when there is no token and 403 when there is a valid token but user is not authorized.

**Comment by maxceem** [comment]
Yep, though don't remember cases when we throw `403` we throw `404` instead as if the record is not found.

**Comment by vikasrohit** [comment]
Yep, that is true as well because we don't want to let the users know if the record really exists and they don't have the access. Further, I think for the particular project listing endpoint, we allow every valid topcoder user (i.e. every valid token) to be able to call the endpoint and just return the projects user is part of.
How the other v5 APIs are handling this right now?

**Comment by vikasrohit** [comment]
Okay. I need some time to think about it, will come back some time later.

**Comment by maxceem** [comment]
One more case to consider raised up on the forum https://apps.topcoder.com/forums/?module=Thread&threadID=949559&mc=9#2386319 during the challenge https://www.topcoder.com/challenges/30112672.

Our current `PATCH /v5/project/{id}/invites/{inviteId}` endpoint returns `403` when a regular user tries to update invite for some other user. Which feel the right thing to do https://github.com/topcoder-platform/tc-project-service/blob/hotfix/project-search-sql/src/routes/projectMemberInvites/update.js#L79-L86

During implementing `GET` endpoint `GET /v5/project/{id}/invites/{inviteId}` I set the requirement to return error `404` if a `non-member` tries to get an invitation for other users. This also feels right as per our logic we return `404` instead of `403` when users access somethings they don't have access to.

Though we end up with inconsistency here, in `GET` endpoint we hide the fact the invite for another user exists by returning `404`. But in `PATCH` endpoint we expose the fact that invitation for another user exists by returning `403`.

**Comment by vikasrohit** [comment]
I guess, we should follow 404 for PATCH invite endpoint to be consistent with our own pattern. We would decide later if we need to throw 403 instead of 404.

**Comment by maxceem** [comment]
There are also some peculiarity in this particular case. We have different reasons to return `403` now https://github.com/topcoder-platform/tc-project-service/blob/hotfix/project-search-sql/src/routes/projectMemberInvites/update.js#L69-L80:
1. `Requested invites can only be updated by Copilot manager`
2. `Project members can cancel invites only for customers`
3. `Project members can only update invites for themselves`

In cases `1` and `2` users can `GET` such invites and **can see** them, but **cannot update** them.
in case `3` users cannot `GET` them and cannot update.

So it makes sense to return `403` in cases 1 and 2 as users **can see** such invites, but **cannot** update. And return `404` as users **cannot see** such invite and cannot update.

Though on the other when some `client` application uses such API, having different error codes in such cases may be unexpected.

**Comment by vikasrohit** [comment]
I think this behaviour is bit driven by the complicated logic required for the invite process. 1 is good for throwing 403, 2 is going away with our upcoming [changes](https://github.com/topcoder-platform/tc-project-service/issues/437#issuecomment-572527231) where we won't allow people to cancel an invite using the PATCH endpoint. 3 is actually the confusing part where the invite does exist but we throw 404. But lest keep it as is for now, we can think of it later. I don't want any feature to be disturbed because of these changes.
